src/arithmetic/algebraic_expression/algebraic_expression_eval.c

```diff
@@ -235,18 +193,25 @@ GrB_Matrix _AlgebraicExpression_Eval(const AlgebraicExpression *exp, GrB_Matrix
 		res = exp->operand.matrix;
 		break;
 	default:
-		assert("Unknow algebraic expression node type" && false);
+		assert("Unknown algebraic expression node type" && false);
 	}
 
 	return res;
 }
 
 void AlgebraicExpression_Eval(const AlgebraicExpression *exp, GrB_Matrix res) {
 	assert(exp && exp->type == AL_OPERATION);
+	_AlgebraicExpression_Eval(exp, res);
+}
 
-	// On first evaluation we need to fetch operands
-	_AlgebraicExpression_FetchOperands((AlgebraicExpression *)exp, QueryCtx_GetGraphCtx(),
-									   QueryCtx_GetGraph());
+void AlgebraicExpression_Initialize(AlgebraicExpression **exp, GrB_Matrix filter_matrix) {
+	// On first evaluation we need to fetch operands.
+	_AlgebraicExpression_FetchOperands(*exp, QueryCtx_GetGraphCtx(), QueryCtx_GetGraph());
 
-	_AlgebraicExpression_Eval(exp, res);
+	// Add placeholder operand to hold filter matrix as leftmost multiplication.
+	AlgebraicExpression_MultiplyToTheLeft(exp, filter_matrix);
+
+	// Optimize the expression tree.
+	AlgebraicExpression_Optimize(exp);
 }
```

Comment:
Please remove this function, it is too specific, Algebraic expression should stay as generic as possible.